sys/event/event.c

```diff
@@ -18,20 +18,21 @@ void event_queue_init(event_queue_t *queue)
 {
     assert(queue);
     memset(queue, '\0', sizeof(*queue));
-    queue->waiter = (thread_t *)sched_active_thread;
 }
 
 void event_post(event_queue_t *queue, event_t *event)
 {
-    assert(queue && queue->waiter && event);
+    assert(queue && event);
 
     unsigned state = irq_disable();
     if (!event->list_node.next) {
         clist_rpush(&queue->event_list, &event->list_node);
     }
-    irq_restore(state);
 
-    thread_flags_set(queue->waiter, THREAD_FLAG_EVENT);
+    if (queue->waiter) {
+        thread_flags_set(queue->waiter, THREAD_FLAG_EVENT);
+    }
+    irq_restore(state);
```

Comment:
Wait... is it okay, that you flipped the order of `irq_restore()` and `thread_flags_set()` here?

Reply:
nope, it isn't

Reply:
IMHO it should be ok: `thread_flags_set()` should work in case it is called with interrupts disabled, so the potential task switch will happen after they are re-enabled in line 35. Having the flags set inside the critical section here prevents any concurrency issues that could arise when the 'queue service thread' is interrupt while at certain points in the `event_wait()` function.